### ISSUE TITLE
Allow template expansion in the markdown

### ISSUE THREAD

**Issue by kristofvanlandschoot** [jandecaluwe/urubu#37]
When moving from jekyll to urubu (which went great, thanks!!!), I found that in urubu my markdown needs to be "pure", i.e. it is not possible to contain {% ... %} markers with templating instructions.

Would it be possible to do this in urubu as well?

I needed this because on my site images were not done with img-tags, but with figure-tags (and had a caption as a parameter). The standard markdown templating engine will always render img-tags.

**Comment by jandecaluwe** [comment]
> When moving from jekyll to urubu

Interesting - why did you migrate?

>  I found that in urubu my markdown needs to be "pure", i.e. it is not possible to contain {% ... %} > > markers with templating instructions.
> Would it be possible to do this in urubu as well?

May be easy and useful. I will look into this during the weekend.

**Comment by jandecaluwe** [comment]
I have experimented with this, but I am not convinced.

Now I would have to escape all template delimiters that should not be processed, e.g. when used in documentation - as I do on the Urubu documentation website.

Also, I don't fully understand why you needed this for the special tags. Why not use those tags directly in the Markdown source code.

I still need to be convinced that this is really worth it.

**Comment by xHN35RQ** [comment]
In Jekyll `{% ... %}` delimiters indicate Liquid tags. [Liquid Markup](http://liquidmarkup.org) support is a unique feature of the Jekyll rendering engine and not a part of the standard Markdown syntax. Jekyll supports using these tags directly within a Markdown document.

@kristofvanlandschoot the easiest option is to convert your images into proper [Markdown img format](https://daringfireball.net/projects/markdown/syntax#img):
```
![Alt text](/path/to/img.jpg)
![Alt text](/path/to/img.jpg "Optional title")

* An exclamation mark: !;
* followed by a set of square brackets, containing the alt attribute text for the image;
* followed by a set of parentheses, containing the URL or path to the image, and an optional title attribute enclosed in double or single quotes.

```
If you want to add captions you can do so using HTML directly in your Markdown. You'll need to style it properly with CSS. See https://stackoverflow.com/questions/10128950/how-to-write-a-caption-under-an-image

It seems easy enough to extend Urubu with a template to support image captions. The template could look for the existence of an optional title attribute enclosed in double or single quotes and render that text within `<figcaption></figcaption>` HTML tags.

**Comment by kristofvanlandschoot** [comment]
I understand the not being convinced part. Not sure I am convinced that it is a good idea either but I don't know the urubu code :).

But, case in point: having images render as <figure> with <figcaption> is, IMO, useful and I see not directly another way using only Markdown. Of course I can get by with CSS (as xHN35RQ remarked) and have my captions show up in another way (and this is what I did in the end), but it still feels like a workaround for a real problem that people might have.

(ps: jan, I migrated because I have an interest in static blog site generators and your argument that it is a pity that most of them are focusing too much on blogposts is something I agree with. it worked for me, so I migrated).

**Comment by jandecaluwe** [comment]
Ok, I have stored away my prototype for this on a feature implementation branch. I am doing this because I don't only see advantages (what about processing overhead - no idea), so I need to see a compelling use case to enable this.

Regarding your particular problem - I still don't follow how template expansion makes the task much easier, compared to directly using `<figure>` and `<figcaption>` tags right in your markdown document. (E.g. I always forget the Markdown image syntax, and I tend to use `<img>` tags directly.  I also don't follow the CSS reasoning as if it was overhead - I assume your templating method eventually also relies on `<figure>` and `<figcaption>` tags, so you need the CSS for those anyway?

Another point is that we could indeed consider to patch the Markdown rendered for stand-alone images, like MultiMarkdown does. Nothing against making use of more modern tags I think.

**Comment by xHN35RQ** [comment]
> Another point is that we could indeed consider to patch the Markdown renderer for stand-alone images, like MultiMarkdown does, and generate `<figure>` and `<figcaption>` from the Markdown syntax. Nothing against making use of more modern tags I think.

This makes the most sense in the long run.

**Comment by jandecaluwe** [comment]
In python-markdown, this would be done by adding a block processor that processes standalone image links in the desired way (not like an inline pattern). Doable, not trivial.

**Comment by kristofvanlandschoot** [comment]
In discussing this I feel we are focusing too much on the <img> primitive. I agree there are good enough workarounds for this in the current implementation.

But the point of having the template expansion within markdown is to have a *general* mechanism for a user to define new primitives. Is there, or will there ever be, a need for a user to describe things within markdown that are not defined in markdown.

Maybe I have a better example. For a website I write we have a style guide. It is a music website, and we write song titles in italics, we write album titles within quotation marks and we write group names in bold. Of course sometimes, within an article, a writer will emphasise some sentences, or write things in an ironic way, for which the writer might use bold or italics or both within the text.

Now should it be possible in Markdown to write something like {% albumtitle: 'Black Star' %} in the markdown? That way I can define primitives on a higher level that allow that if I ever change the style guide, and we decide to write album titles in italics instead of within quotes, I only have one place that I have to change instead of going over all of the Markdown files on my site to change the style where appropriate.

In other words, having the possibility of template expansion within markdown lifts the markdown to a level where I can, if I want to, define primitives that are broader than what markdown provides. I think that is a valuable thing to have.

**Comment by xHN35RQ** [comment]
> In python-markdown, this would be done by adding a block processor that processes standalone image links in the desired way (not like an inline pattern). Doable, not trivial.

Yeah, the easiest way is to just use the raw HTML `<figcaption>` tags directly in the Markdown. Which is perfectly valid Markdown.

> Now should it be possible in Markdown to write something like {% albumtitle: 'Black Star' %} in the markdown? 

No, that should not be possible in Markdown. Only Markdown markup and HTML markup is allowable in Markdown.

This might be a feature of Urubu, as a feature of how it interprets Markdown documents, but I think you're probably straying too far out of the stated intention of Urubu and might be better served using another static site generator.

For example, https://krasimir.github.io/techy has built-in support for custom data. It's very much what you're looking for. Hugo offers similar functionality: http://gohugo.io/extras/datafiles/

**Comment by kristofvanlandschoot** [comment]
Interesting. I did not know about techy or hugo, I'll look into them.

**Comment by jandecaluwe** [comment]
@kristofvanlandschoot I start to see the point ;-)

I have experimented a little. I think the syntax you suggest is based on new tags in Liquid. Defining new tags is possible with Jinja2, but it seems rather complex. However, we have macro's instead.

For example, I can define a `figure` macro in file `macros.html` in the `_layout` directory as follows:

```
{% macro figure(fn, caption='') %}
<figure>
  <img src="/img/{{fn}}" class="img-responsive" alt="{{caption}}">
  {% if caption %}
  <figcaption>{{caption}}</figcaption>
  {% endif %}
</figure>
{% endmacro %}
```
and then use it in any Markdown page as follows:

```
{% from 'macros.html' import figure %}
...
{{figure('urubu.jpg', "An Urubu - a brazilian vulture")}}
```
This results in the desired html - I tried this and it worked fine. Note that the `import` knows where to look for `macros.html` because the `_layouts` dir is specified as a search dir in Urubu's template environment. 

Is this what you are looking for? If so, and if you or someone else wants to evaluate this further, I am prepared to push this feature and some documentation ot the public master branch.

**Comment by kristofvanlandschoot** [comment]
This is exactly what I am looking for. I can give it a try and give some feedback, yes.

**Comment by jandecaluwe** [comment]
I have pushed a draft implementation. Documentation is here:

http://urubu.jandecaluwe.com/manual/templating_in_pages.html

The code for these examples is in `doc/_layout/util.html`.

**Comment by xHN35RQ** [comment]
Wow this is impressive. I'll be using this functionality in future. thanks @jandecaluwe !

**Comment by xHN35RQ** [comment]
I noticed `{% raw %}` and `{% endraw %}` being used to surround some code, however there's no mention of them in the documentation.

**Comment by jandecaluwe** [comment]
The variables / macros are actually expanded in the Markdown source first. 

I have not added a section about variables/macros first, because I think that defies the purpose of the examples (diving in immediately) and would suggest that template support is limited to these concepts. However, I have added some conceptual he documentation after the examples, pointing out that the examples merely scratch the surface and that the Jinja2 template documentation is the reference for what can be done. I also added a section about delimiters and escaping because that is important even for simple usage.

**Comment by xHN35RQ** [comment]
Wow, I didn't realize it wasn't limited to just variables and macros. This make so much more sense now, thank you for the added documentation and links to the Jinja documentation. Very helpful.

**Comment by jandecaluwe** [comment]
Does anyone have further feedback on this? I would like to proceed with a release.

**Comment by kristofvanlandschoot** [comment]
I haven't gotten around to testing it yet, sorry. Next Thursday I can make some time.

**Comment by jandecaluwe** [comment]
Thx, that would be nice.

**Comment by kristofvanlandschoot** [comment]
Tested and works for me!

**Comment by kristofvanlandschoot** [comment]
Great job, BTW!

**Comment by jandecaluwe** [comment]
Thx, has been released now in 1.2.0.
